src/main/java/org/simulator/io/CSVImporter.java

```diff
@@ -58,114 +59,109 @@
-        String colNames[] = new String[newHead.length];
+        String[] colNames = new String[newHead.length];
         UniqueNamedSBase sbase;
-        j = 0;
+        int j = 0;
         for (String head : newHead) {
           sbase = model.findUniqueNamedSBase(head);
-          colNames[j++] =
-              (sbase != null) && (sbase.isSetName()) ? sbase.getName() : null;
+          colNames[j++] = (sbase != null) && (sbase.isSetName()) ? sbase.getName() : null;
         }
         data.getBlock(0).setColumnNames(colNames);
       }
       data.setTimeName("time");
+
       return data;
-    } else {
-      logger.fine("The file is not correctly formatted!");
+    } catch (Exception e) {
+      e.printStackTrace();
     }
     return null;
   }
 
   /**
-   * @param pathname
-   * @return
+   * Stores the content from the CSV file in the Hashmap with keys as identifiers
+   * and values as the column of the identifier's amounts.
+   *
+   * @param model the SBML {@link Model}
+   * @param pathname path of the CSV file
+   * @return the results from the CSV file in MultiTable
    * @throws IOException
    */
-  private String[][] read(String pathname) throws IOException {
-    String[][] result = null;
+  public MultiTable readDataFromCSV(Model model, String pathname) throws IOException {
+    Map<String, double[]> identifierToColumn = new HashMap<>();
     BufferedReader reader = new BufferedReader(new FileReader(pathname));
-    List<String> lines = new LinkedList<String>();
     String line = reader.readLine();
+    List<String> lines = new LinkedList<String>();
+    header = line.split(",");
     if (line != null) {
-      header = line.split(",");
       line = reader.readLine();
       while ((line != null) && !line.isEmpty()) {
         lines.add(line);
         line = reader.readLine();
       }
-      result = new String[lines.size()][header.length];
-      int i = 0;
-      for (String l : lines) {
-        result[i] = l.split(",");
-        i++;
+
+      for (String s : header) {
+        identifierToColumn.put(s, new double[lines.size()]);
+      }
+
+      for (int i = 0; i < lines.size(); i++) {
+        String[] column = lines.get(i).split(",");
+        for (int j = 0; j < column.length; j++) {
+          if ((column[j] != null) && (column[j].length() > 0)) {
+            if (column[j].equalsIgnoreCase("INF")) {
+              identifierToColumn.get(header[j])[i] = Double.POSITIVE_INFINITY;
+            } else if (column[j].equalsIgnoreCase("-INF")) {
+              identifierToColumn.get(header[j])[i] = Double.NEGATIVE_INFINITY;
+            } else if (column[j].equalsIgnoreCase("NAN")) {
+              identifierToColumn.get(header[j])[i] = Double.NaN;
+            } else {
+              identifierToColumn.get(header[j])[i] = Double.parseDouble(column[j]);
+            }
+          }
+        }
       }
     }
-    reader.close();
-    return result;
+
+    return adaptDataToModel(model, identifierToColumn);
```

Comment:
This still mixes all the functions together. 
1. The `readDataFromCSV`should return the HashMap (`Map<String, double[]> identifierToColumn = new HashMap<>();`). Do not call the adaptDataToModel function in the readDataFunction. Also call the hashMap just `columns` or `columnsMap`. Remove the header instance variable and the side effects by it! The header keys are the keys in the HashMap, no need for an instance variable here. You can preserve the order by using a LinkedHashMap (see https://stackoverflow.com/questions/10710193/how-to-preserve-insertion-order-in-hashmap).

2. Add a second function readMultitableFromCSV which returns a MultiTable and uses the readData function. It can have an optional model instance. This is the `adaptDataToModel`. 

Unfortunately I still don't understand what the function is doing exactly. Can you describe this in 1-2 sentences? I am very sure the current implementation is much too complicated for what the function should do. But unfortunately I have no clue what the function should do.

Reply:
@matthiaskoenig, I have updated the things you requested. 

In the latest commit, 
readDataFromCSV():
This method takes a path and reads each of the columns from the CSV files and stores them in a LinkedListHashMap with the key as `column name` and value as `an array with its corresponding amounts`.

readMultiTableFromCSV():
In this method, even I am quite confused as I don't see any need of model here. I have taken that model conditions directly from the previous implementation and have copied it but I don't see it of any use (but haven't removed it yet as I am not sure whether it has some use case or not). Other than model usage, I have now taken the map from the readDataFromCSV() method and converted it to the MultiTable.

Also with this, I have removed the global `header` array.